File: common/src/main/java/maestro/api/schematic/MirroredSchematic.java
package maestro.api.schematic;

import java.util.List;
import java.util.stream.Collectors;
import net.minecraft.world.level.block.Mirror;
import net.minecraft.world.level.block.state.BlockState;

public class MirroredSchematic implements ISchematic {

    private final ISchematic schematic;
    private final Mirror mirror;

    public MirroredSchematic(ISchematic schematic, Mirror mirror) {
        this.schematic = schematic;
        this.mirror = mirror;
    }

    @Override
    public boolean inSchematic(int x, int y, int z, BlockState currentState) {
        return schematic.inSchematic(
                mirrorX(x, widthX(), mirror),
                y,
                mirrorZ(z, lengthZ(), mirror),
                mirror(currentState, mirror));
    }

    @Override
    public BlockState desiredState(
            int x, int y, int z, BlockState current, List<BlockState> approxPlaceable) {
        return mirror(
                schematic.desiredState(
                        mirrorX(x, widthX(), mirror),
                        y,
                        mirrorZ(z, lengthZ(), mirror),
                        mirror(current, mirror),
                        mirror(approxPlaceable, mirror)),
                mirror);
    }

    @Override
    public void reset() {
        schematic.reset();
    }

    @Override
    public int widthX() {
        return schematic.widthX();
    }

    @Override
    public int heightY() {
        return schematic.heightY();
    }

    @Override
    public int lengthZ() {
        return schematic.lengthZ();
    }

    private static int mirrorX(int x, int sizeX, Mirror mirror) {
        return switch (mirror) {
            case NONE, LEFT_RIGHT -> x;
            case FRONT_BACK -> sizeX - x - 1;
        };
    }

    private static int mirrorZ(int z, int sizeZ, Mirror mirror) {
        return switch (mirror) {
            case NONE, FRONT_BACK -> z;
            case LEFT_RIGHT -> sizeZ - z - 1;
        };
    }

    private static BlockState mirror(BlockState state, Mirror mirror) {
        if (state == null) {
            return null;
        }
        return state.mirror(mirror);
    }

    private static List<BlockState> mirror(List<BlockState> states, Mirror mirror) {
        if (states == null) {
            return null;
        }
        return states.stream().map(s -> mirror(s, mirror)).collect(Collectors.toList());
    }
}

File: common/src/main/maestro/schematic/MirroredSchematic.kt
package maestro.schematic

import net.minecraft.world.level.block.Mirror
import net.minecraft.world.level.block.state.BlockState

class MirroredSchematic(
    schematic: ISchematic,
    private val mirror: Mirror,
) : TransformedSchematic(schematic) {
    override fun inSchematic(
        x: Int,
        y: Int,
        z: Int,
        currentState: BlockState?,
    ): Boolean =
        schematic.inSchematic(
            mirrorX(x, widthX(), mirror),
            y,
            mirrorZ(z, lengthZ(), mirror),
            mirror(currentState, mirror),
        )

    override fun desiredState(
        x: Int,
        y: Int,
        z: Int,
        current: BlockState?,
        approxPlaceable: List<BlockState>?,
    ): BlockState =
        mirror(
            schematic.desiredState(
                mirrorX(x, widthX(), mirror),
                y,
                mirrorZ(z, lengthZ(), mirror),
                mirror(current, mirror),
                mirror(approxPlaceable, mirror),
            ),
            mirror,
        )!!

    companion object {
        private fun mirrorX(
            x: Int,
            sizeX: Int,
            mirror: Mirror,
        ): Int =
            when (mirror) {
                Mirror.NONE, Mirror.LEFT_RIGHT -> x
                Mirror.FRONT_BACK -> sizeX - x - 1
            }

        private fun mirrorZ(
            z: Int,
            sizeZ: Int,
            mirror: Mirror,
        ): Int =
            when (mirror) {
                Mirror.NONE, Mirror.FRONT_BACK -> z
                Mirror.LEFT_RIGHT -> sizeZ - z - 1
            }

        private fun mirror(
            state: BlockState?,
            mirror: Mirror,
        ): BlockState? = state?.mirror(mirror)

        private fun mirror(
            states: List<BlockState>?,
            mirror: Mirror,
        ): List<BlockState>? = states?.map { mirror(it, mirror)!! }
    }
}
